### ISSUE TITLE
A3C: state_processor.py

### ISSUE THREAD

**Issue by IbrahimSobh** [dennybritz/reinforcement-learning#42]
Hi

I wanted to conduct some extra pre-processing for the state (games image). And see how this may speed up the training.

For example, in Break-out, I thought that the most important part for the agent (specially at the beginning of the training) is to neglect the wall at all and just concentrate on the ball and try to hit it back (This is not the best assumption, just an example).

So, I decided to **remove the wall** as in the following image:
![no_wall](https://cloud.githubusercontent.com/assets/19908396/20898903/cfb56e24-bb30-11e6-95df-46f4e31915ec.png)

In state_process.py, I made this minor modification:

> def process(self, state, sess=None):
>         **state[30:95] = 0**
>         sess = sess or tf.get_default_session()
>         return sess.run(self.output, { self.input_state: state })
> 

My questions are: 

1. Should I conduct any pre-processing in "process" function as above, or in --init--(self) and use tf.image? and why?
2. How to use tf.image to set state[30:95] = 0 
3. If I want to re-shape the reward by giving a small reward to the agent if the bat hits the ball, Where to do this in the code? I think I need to detect myself if the ball hits the bat using some processing of the state, and then change the reward.  

Thank you

**Comment by Shadowen** [comment]
These are all very good points and it does seem that you can improve the performance of the agent. However, the original intent of the A3C algorithm (as well as DQN and all its predecessors) is to create one algorithm that can play a variety of games. In this case, the StateProcessor is designed for all Atari games. 

Your custom feature engineering works fine for Breakout, but does it work for all Atari games without manual intervention?

**Comment by IbrahimSobh** [comment]
Thank you

Exactly, I want to use specific game knowledge to enhance the training process. If it works, then I will think how to "guess" the game knowledge implicitly.


The problem now is: Performance of DQN  (and A3C) is not compatible with the papers!  Kindly see 
https://github.com/dennybritz/reinforcement-learning/issues/30

**Comment by waterzxj** [comment]
------------------ 原始邮件 ------------------
发件人: "Ibrahim Sobh";<notifications@github.com>;
发送时间: 2017年3月9日(星期四) 晚上11:25
收件人: "dennybritz/reinforcement-learning"<reinforcement-learning@noreply.github.com>; 
抄送: "Subscribed"<subscribed@noreply.github.com>; 
主题: Re: [dennybritz/reinforcement-learning] A3C: state_processor.py (#42)




Thank you
 
Exactly, I want to use specific game knowledge to enhance the training process. If it works, then I will think how to "guess" the game knowledge implicitly.
 
The problem now is: Performance of DQN  (and A3C) is not compatible with the papers!  Kindly see
 #30

**Comment by LinZichuan** [comment]
Hi, feed_dict in Tensorflow seems very slow. Using sess.run to process each state seems not a good idea.

**Comment by Shadowen** [comment]
How else do you... use Tensorflow?

**Comment by LinZichuan** [comment]
Maybe it does not need to use Tensorflow to process each state. Using numpy or other libraries can avoid use feed_dict frequently.
